### PR TITLE
Example of using asyncio and tornadoweb together with pyzmq

### PR DESCRIPTION
**Pull request by RojavaCrypto**

### CODE REVIEW

**Comment by RojavaCrypto** [comment]
Hmm, I just added ZMQ sockets to the example but now it's crashing with:

```
$ python3 tornado_asyncio.py 
ERROR:tornado.application:Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x7f0187db8f28>, <tornado.concurrent.Future object at 0x7f01867983c8>)
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/tornado/ioloop.py", line 600, in _run_callback
    ret = callback()
  File "/usr/local/lib/python3.5/dist-packages/tornado/stack_context.py", line 275, in null_wrapper
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tornado/ioloop.py", line 615, in <lambda>
    self.add_future(ret, lambda f: f.result())
  File "/usr/local/lib/python3.5/dist-packages/tornado/concurrent.py", line 232, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 3, in raise_exc_info
  File "/usr/local/lib/python3.5/dist-packages/tornado/gen.py", line 282, in wrapper
    yielded = next(result)
  File "<string>", line 6, in _wrap_awaitable
  File "tornado_asyncio.py", line 14, in pushing
    await server.send(b"Hello")
  File "/usr/local/lib/python3.5/dist-packages/zmq/eventloop/future.py", line 164, in send
    args=dict(flags=flags, copy=copy, track=track),
  File "/usr/local/lib/python3.5/dist-packages/zmq/eventloop/future.py", line 208, in _add_send_event
    self._add_io_state(self._WRITE)
  File "/usr/local/lib/python3.5/dist-packages/zmq/asyncio.py", line 285, in _add_io_state
    self.io_loop.add_writer(self, self._handle_send)
  File "/usr/lib/python3.5/asyncio/selector_events.py", line 272, in add_writer
    key = self._selector.get_key(fd)
  File "/usr/lib/python3.5/selectors.py", line 189, in get_key
    return mapping[fileobj]
  File "/usr/lib/python3.5/selectors.py", line 70, in __getitem__
    fd = self._selector._fileobj_lookup(fileobj)
  File "/usr/lib/python3.5/selectors.py", line 224, in _fileobj_lookup
    return _fileobj_to_fd(fileobj)
  File "/usr/lib/python3.5/selectors.py", line 39, in _fileobj_to_fd
    "{!r}".format(fileobj)) from None
ValueError: Invalid file object: <zmq.asyncio.Socket object at 0x7f0186790c48>
ERROR:tornado.application:Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x7f0187dc50d0>, <tornado.concurrent.Future object at 0x7f01867a7cf8>)
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/tornado/ioloop.py", line 600, in _run_callback
    ret = callback()
  File "/usr/local/lib/python3.5/dist-packages/tornado/stack_context.py", line 275, in null_wrapper
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tornado/ioloop.py", line 615, in <lambda>
    self.add_future(ret, lambda f: f.result())
  File "/usr/local/lib/python3.5/dist-packages/tornado/concurrent.py", line 232, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 3, in raise_exc_info
  File "/usr/local/lib/python3.5/dist-packages/tornado/gen.py", line 282, in wrapper
    yielded = next(result)
  File "<string>", line 6, in _wrap_awaitable
  File "tornado_asyncio.py", line 20, in pulling
    greeting = await client.recv()
  File "/usr/local/lib/python3.5/dist-packages/zmq/eventloop/future.py", line 144, in recv
    dict(flags=flags, copy=copy, track=track)
  File "/usr/local/lib/python3.5/dist-packages/zmq/eventloop/future.py", line 199, in _add_recv_event
    self._add_io_state(self._READ)
  File "/usr/local/lib/python3.5/dist-packages/zmq/asyncio.py", line 283, in _add_io_state
    self.io_loop.add_reader(self, self._handle_recv)
  File "/usr/lib/python3.5/asyncio/selector_events.py", line 234, in add_reader
    key = self._selector.get_key(fd)
  File "/usr/lib/python3.5/selectors.py", line 189, in get_key
    return mapping[fileobj]
  File "/usr/lib/python3.5/selectors.py", line 70, in __getitem__
    fd = self._selector._fileobj_lookup(fileobj)
  File "/usr/lib/python3.5/selectors.py", line 224, in _fileobj_lookup
    return _fileobj_to_fd(fileobj)
  File "/usr/lib/python3.5/selectors.py", line 39, in _fileobj_to_fd
    "{!r}".format(fileobj)) from None
ValueError: Invalid file object: <zmq.asyncio.Socket object at 0x7f0186790b88>
^CTraceback (most recent call last):
  File "tornado_asyncio.py", line 31, in <module>
    zmq_tornado_loop()
  File "tornado_asyncio.py", line 28, in zmq_tornado_loop
    loop.start()
  File "/usr/local/lib/python3.5/dist-packages/tornado/platform/asyncio.py", line 121, in start
    self.asyncio_loop.run_forever()
  File "/usr/lib/python3.5/asyncio/base_events.py", line 331, in run_forever
    self._run_once()
  File "/usr/lib/python3.5/asyncio/base_events.py", line 1262, in _run_once
    event_list = self._selector.select(timeout)
  File "/usr/lib/python3.5/selectors.py", line 441, in select
    fd_event_list = self._epoll.poll(timeout, max_ev)
KeyboardInterrupt
```

Any ideas?

File: examples/asyncio/tornado_asyncio.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import asyncio
+import zmq.asyncio
+from tornado.platform.asyncio import AsyncIOMainLoop
+AsyncIOMainLoop().install()
+
+# This must be instantiated after the installing the IOLoop
```
Comment (minrk) [inline examples/asyncio/tornado_asyncio.py]: I think you still need to do `zmq.asyncio.install()` to register pyzmq as the asyncio eventloop.

**Comment by RojavaCrypto** [comment]
If I use that it still doesn't work (just hangs). Here's the code for your testing:

```python
import asyncio
import zmq.asyncio
#from tornado.platform.asyncio import AsyncIOMainLoop
#AsyncIOMainLoop().install()
zmq.asyncio.install()

# This must be instantiated after the installing the IOLoop
ctx = zmq.asyncio.Context()

async def pushing():
    server = ctx.socket(zmq.PUSH)
    server.bind('tcp://*:9000')
    while True:
        await server.send(b"Hello")

async def pulling():
    client = ctx.socket(zmq.PULL)
    client.connect('tcp://127.0.0.1:9000')
    while True:
        greeting = await client.recv()
        print(greeting)

def zmq_tornado_loop():
    from zmq.eventloop.ioloop import IOLoop
    loop = IOLoop.current()
    loop.spawn_callback(pushing)
    loop.spawn_callback(pulling)
    loop.start()

if __name__ == '__main__':
    zmq_tornado_loop()

```

**Comment by achimnol** [comment]
I have encountered the same problem.

After some introspection, I found that `_add_io_state()` method of the `zmq.asyncio.Socket` class seems to pass the Socket object itself to `add_reader()` / `add_writer()` instead of the underlying file descriptor. Adding an extra call `fd = self.get(_zmq.FD)` and replacing the first arguments seems to resolve the exception.

```python
    def _add_io_state(self, state):
        """Add io_state to poller."""
        if not self._state & state:
            self._state = self._state | state
            fd = self.get(_zmq.FD)
            if state & self._READ:
                self.io_loop.add_reader(fd, self._handle_recv)
            if state & self._WRITE:
                self.io_loop.add_writer(fd, self._handle_send)

    def _drop_io_state(self, state):
        """Stop poller from watching an io_state."""
        if self._state & state:
            self._state = self._state & (~state)
            fd = self.get(_zmq.FD)
            if state & self._READ:
                self.io_loop.remove_reader(fd)
            if state & self._WRITE:
                self.io_loop.remove_writer(fd)
```

However, still the code modified from above (to match with asyncio's API) does not work:

```python
import asyncio
import zmq.asyncio
zmq.asyncio.install() # does not matter...

ctx = zmq.asyncio.Context()

async def pushing():
    server = ctx.socket(zmq.PUSH)
    server.bind('tcp://*:9000')
    asyncio.sleep(0.5)
    while True:
        await server.send(b"Hello")

async def pulling():
    client = ctx.socket(zmq.PULL)
    client.connect('tcp://127.0.0.1:9000')
    asyncio.sleep(1)
    while True:
        greeting = await client.recv()
        print(greeting)

def main():
    loop = asyncio.get_event_loop()
    loop.create_task(pushing())
    loop.create_task(pulling())
    loop.run_forever()
    loop.close()

if __name__ == '__main__':
    main()
```

```
Task exception was never retrieved
future: <Task finished coro=<pushing() done, defined at test.py:7> exception=Again('Resource temporarily unavailable')>
Traceback (most recent call last):
  File "/usr/local/opt/pyenv/versions/3.5.1/lib/python3.5/asyncio/tasks.py", line 241, in _step
    result = coro.throw(exc)
  File "test.py", line 12, in pushing
    await server.send(b"Hello")
  File "/usr/local/opt/pyenv/versions/3.5.1/lib/python3.5/asyncio/futures.py", line 358, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/local/opt/pyenv/versions/3.5.1/lib/python3.5/asyncio/tasks.py", line 290, in _wakeup
    future.result()
  File "/usr/local/opt/pyenv/versions/3.5.1/lib/python3.5/asyncio/futures.py", line 274, in result
    raise self._exception
  File "/usr/local/opt/pyenv/versions/3.5.1/lib/python3.5/site-packages/zmq/eventloop/future.py", line 298, in _handle_send
    result = send(msg, **kwargs)
  File "zmq/backend/cython/socket.pyx", line 636, in zmq.backend.cython.socket.Socket.send (zmq/backend/cython/socket.c:6946)
  File "zmq/backend/cython/socket.pyx", line 683, in zmq.backend.cython.socket.Socket.send (zmq/backend/cython/socket.c:6690)
  File "zmq/backend/cython/socket.pyx", line 206, in zmq.backend.cython.socket._send_copy (zmq/backend/cython/socket.c:2721)
  File "zmq/backend/cython/socket.pyx", line 201, in zmq.backend.cython.socket._send_copy (zmq/backend/cython/socket.c:2609)
  File "zmq/backend/cython/checkrc.pxd", line 19, in zmq.backend.cython.checkrc._check_rc (zmq/backend/cython/socket.c:7684)
zmq.error.Again: Resource temporarily unavailable
Task exception was never retrieved
future: <Task finished coro=<pulling() done, defined at test.py:14> exception=Again('Resource temporarily unavailable')>
Traceback (most recent call last):
  File "/usr/local/opt/pyenv/versions/3.5.1/lib/python3.5/asyncio/tasks.py", line 241, in _step
    result = coro.throw(exc)
  File "test.py", line 19, in pulling
    greeting = await client.recv()
  File "/usr/local/opt/pyenv/versions/3.5.1/lib/python3.5/asyncio/futures.py", line 358, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/local/opt/pyenv/versions/3.5.1/lib/python3.5/asyncio/tasks.py", line 290, in _wakeup
    future.result()
  File "/usr/local/opt/pyenv/versions/3.5.1/lib/python3.5/asyncio/futures.py", line 274, in result
    raise self._exception
  File "/usr/local/opt/pyenv/versions/3.5.1/lib/python3.5/site-packages/zmq/eventloop/future.py", line 263, in _handle_recv
    result = recv(**kwargs)
  File "zmq/backend/cython/socket.pyx", line 693, in zmq.backend.cython.socket.Socket.recv (zmq/backend/cython/socket.c:7283)
  File "zmq/backend/cython/socket.pyx", line 727, in zmq.backend.cython.socket.Socket.recv (zmq/backend/cython/socket.c:7081)
  File "zmq/backend/cython/socket.pyx", line 150, in zmq.backend.cython.socket._recv_copy (zmq/backend/cython/socket.c:2126)
  File "zmq/backend/cython/socket.pyx", line 145, in zmq.backend.cython.socket._recv_copy (zmq/backend/cython/socket.c:2033)
  File "zmq/backend/cython/checkrc.pxd", line 19, in zmq.backend.cython.checkrc._check_rc (zmq/backend/cython/socket.c:7684)
zmq.error.Again: Resource temporarily unavailable
```

**Comment by minrk** [comment]
@achimnol using zmq.FD would not be correct. This is not the underlying FD of the socket, but an edge-triggered FD-like API for signaling zmq events (of any kind) on a socket, and quite difficult to use reliably, and one that I wouldn't recommend anyone use unless absolutely necessary. pyzmq doesn't use this for eventloop integration, instead it uses an implementation of the inner loop that uses `zmq_poll`, so that zmq sockets can be monitored directly, rather than through this extra API.

@RojavaCrypto your example isn't working because the right eventloops aren't running:

1. asyncio hasn't been told to use zmq
2. tornado is told to use asyncio, then told again to use zmq instead of asyncio

This patch should get the right stack of things:

1. tell asyncio to use zmq
2. tell tornado to use asyncio

```diff
diff --git a/examples/asyncio/tornado_asyncio.py b/examples/asyncio/tornado_asyncio.py
index 83f562f..3f707b0 100644
--- a/examples/asyncio/tornado_asyncio.py
+++ b/examples/asyncio/tornado_asyncio.py
@@ -1,6 +1,12 @@
 import asyncio
 import zmq.asyncio
+
+from tornado.ioloop import IOLoop
 from tornado.platform.asyncio import AsyncIOMainLoop
+
+# tell asyncio to use zmq's eventloop
+zmq.asyncio.install()
+# tell tornado to use asyncio
 AsyncIOMainLoop().install()
 
 # This must be instantiated after the installing the IOLoop
@@ -12,6 +18,7 @@ async def pushing():
     server.bind('tcp://*:9000')
     while True:
         await server.send(b"Hello")
+        await asyncio.sleep(1)
 
 async def pulling():
     client = ctx.socket(zmq.PULL)
@@ -21,7 +28,6 @@ async def pulling():
         print(greeting)
 
 def zmq_tornado_loop():
-    from zmq.eventloop.ioloop import IOLoop
     loop = IOLoop.current()
     loop.spawn_callback(pushing)
     loop.spawn_callback(pulling)
```

**Comment by RojavaCrypto** [comment]
Just tested those changes in the example and my main project. It works correctly. I've updated the example.

**Comment by minrk** [comment]
Thanks!

**Comment by achimnol** [comment]
@minrk Thanks for clarification. Could you show an example not depending on tornado but only on asyncio?
I would like to replace aiozmq with the vanilla pyzmq for sake of reduced denpendencies. aiozmq works fine with the vanilla asyncio event loops as well as uvloop (a faster libuv-based asyncio event loop implementation).

**Comment by achimnol** [comment]
I have found that the current asyncio-only example relies on an explicit poller and ZMQEventLoop. (I was on the subway and didn't check the examples directory when writing above comment. Sorry.)
The example works well, but I wonder whether it is still possible to replace the event loop implementation with another.

**Pull request merged**